### PR TITLE
add hiera support using create_resources

### PR DESCRIPTION
**Pull request by tobru**
With this addition it will be possible to use hiera directly to create repositories or mirrors. Example:
```
aptly::aptly_repos:
  myrepo:
    component: universe
```
The `require => Class['aptly']` had to be removed because this creates a circular dependency. This is also not needed anymore because when using the `aptly_*` parameters of the main class, you already have this dependency satisfied.

Another small addition is to create the `rootDir` and the `public`subfolder if it is configued.

### CODE REVIEW

**Comment by rjw1** [comment]
Thanks for the contribution. One of us will hopefully look at this soon. 

If you could look at the failing tests in the mean time that would be helpful

**Comment by tobru** [comment]
I ran the tests locally and they are happy:
```
% ruby --version      
ruby 1.9.3p551 (2014-11-13 revision 48407) [x86_64-linux]
% rvm --version
rvm 1.26.9 (latest) by Wayne E. Seguin <wayneeseguin@gmail.com>, Michal Papis <mpapis@gmail.com> [https://rvm.io/]
% bundle --version
Bundler version 1.7.11
% gem --version
2.4.5
% rm Gemfile.lock || true
% bundle install --jobs=3 --retry=3

Fetching gem metadata from https://rubygems.org/...........
Resolving dependencies...
Installing CFPropertyList 2.3.0
Installing rake 10.4.2
Installing builder 3.2.2
Installing diff-lcs 1.2.5
Installing excon 0.43.0
Installing facter 1.7.6
Installing formatador 0.2.5
Installing mime-types 1.25.1
Installing net-ssh 2.9.2
Installing multi_json 1.10.1
Installing ipaddress 0.8.0
Installing inflecto 0.0.2
Installing json_pure 1.8.2
Installing nokogiri 1.5.11
Installing kwalify 0.7.2
Installing metaclass 0.0.4
Installing rgen 0.6.6
Installing puppet-lint 1.1.0
Installing rspec-core 2.99.2
Installing rspec-mocks 2.99.3
Installing trollop 2.1.1
Using bundler 1.7.11
Installing systemu 2.6.4
Installing fission 0.5.0
Installing puppet-syntax 1.4.1
Installing rspec-expectations 2.99.2
Installing net-scp 1.2.1
Installing fog-json 1.0.0
Installing hiera 1.3.4
Installing mocha 1.1.0
Installing rbvmomi 1.8.2
Installing rspec 2.99.0
Installing fog-core 1.27.3
Installing rspec-puppet 1.0.1
Installing puppet 3.2.4
Installing fog-xml 0.1.1
Installing fog-sakuracloud 0.1.2
Installing fog-brightbox 0.7.1
Installing fog-serverlove 0.1.1
Installing fog-softlayer 0.3.30
Installing fog-storm_on_demand 0.1.0
Installing fog-vmfusion 0.0.1
Installing puppetlabs_spec_helper 0.8.2
Installing fog-atmos 0.1.0
Installing fog-ecloud 0.0.2
Installing fog-aws 0.0.6
Installing fog-profitbricks 0.0.1
Installing fog-radosgw 0.0.3
Installing fog-terremark 0.0.3
Installing fog-voxel 0.0.2
Installing fog 1.27.0
Installing rspec-system 2.8.0
Installing rspec-system-puppet 2.2.1
Your bundle is complete!
Use `bundle show [gemname]` to see where a bundled gem is installed.
% bundle exec rake test
---> syntax:manifests
---> syntax:templates
---> syntax:hiera:yaml
Cloning into 'spec/fixtures/modules/stdlib'...
remote: Counting objects: 6183, done.
remote: Compressing objects: 100% (28/28), done.
remote: Total 6183 (delta 8), reused 2 (delta 0)
Receiving objects: 100% (6183/6183), 1.23 MiB | 1.06 MiB/s, done.
Resolving deltas: 100% (2637/2637), done.
Checking connectivity... done.
Cloning into 'spec/fixtures/modules/apt'...
remote: Counting objects: 2936, done.
remote: Compressing objects: 100% (1323/1323), done.
remote: Total 2936 (delta 1552), reused 2915 (delta 1545)
Receiving objects: 100% (2936/2936), 626.34 KiB | 565.00 KiB/s, done.
Resolving deltas: 100% (1552/1552), done.
Checking connectivity... done.
/home/tobru/.rvm/rubies/ruby-1.9.3-p551/bin/ruby -S rspec spec/classes/init_spec.rb spec/defines/mirror_spec.rb spec/defines/repo_spec.rb --color
.......................

Finished in 2.7 seconds
23 examples, 0 failures
```
There must be something wrong with Travis =(

**Comment by dcarley** [comment]
Thanks, this is a good feature!

Quick stylistic point - could you apply our [contributing guidelines](https://github.com/gds-operations/puppet-aptly/blob/master/CONTRIBUTING.md#git-workflow) with regards to commit messages?

- You've included some good detail in the pull request description that would be preferable to have in the commit message itself.
- The directory changes aren't related to the other feature so should be broken out to a separate commit.

> The require => Class['aptly'] had to be removed because this creates a circular dependency. This is also not needed anymore because when using the aptly_* parameters of the main class, you already have this dependency satisfied.

These changes are what appear to be making the tests fail. They fail for me locally, but I'm not sure why you get a different result. I think we want to keep these dependencies because they make sure that Puppet has installed and configured aptly prior to creating any repos or mirrors. Without them there is no guarantee what order Puppet will execute the resources in.

Would you be able to add some new rspec tests to confirm that these new params are working correctly? This would also demonstrate any circular dependencies and then we can work out how to fix them.

> Another small addition is to create the rootDir and the publicsubfolder if it is configued.

I think this is normally done by `aptly publish`. Is there any benefit to doing it in Puppet? I think it may cause us problems because we manage the mount/directory as a resource elsewhere in our Puppet codebase.

File: manifests/init.pp [hunk 1]
```diff
@@ -27,15 +27,29 @@
...
+#   Hash of aptly mirrors which is passed to aptly::mirror
+#   Default: {}
+#
 class aptly (
   $package_ensure = present,
   $config = {},
   $repo = true,
   $key_server = undef,
   $user = 'root',
+  ### START Hiera Lookups ###
```
Comment (dcarley) [inline manifests/init.pp]: I'm not sure these comments are strictly necessary because any of the parameters can be provided by hiera lookups.

**Comment by tobru** [comment]
Thanks for your feedback. I'll look into it ASAP

**Comment by rjw1** [comment]
Hi @tobru,

Have you had a chance to look at this?

**Comment by tobru** [comment]
I've reverted changes which don't belong into this PR and did some cleanup. The problem is: I don't know rspec yet and I'm not able to add a test for this new parameters.
When you now wan't to add a aptly repo using the `aptly_repos` parameter, it ends in a dependency cycle:
```
Error: Could not apply complete catalog: Found 1 dependency cycle:
(Exec[aptly_repo_create-myrepo] => Aptly::Repo[myrepo] => Class[Aptly] => Exec[aptly_repo_create-myrepo])
```
At the moment I don't have time to look deeper into it. But it's on my list and as soon as I get some more time, I will look into it again. If someone from you finds some time, this would also be great.

**Comment by ajlanghorn** [comment]
@alexmuller has fixed the tests, by realising that the issue was that test frameworks were pinned to specific versions rather than accepting the latest. This meant that tests were failing on the master branch. He's fixed these now, so if you rebase this PR against master, you should get some working tests.

We've been using this syntax in Hiera to test this change inside our Puppet repository:

```
aptly::aptly_repos:
  my_example_repo:
    component: foo
```

This seems to work fine, and we can now replicate your dependency cycle issue. We're going to look at the code in a bit more depth now and see if we can work out where the cycle occurs.

**Comment by ajlanghorn** [comment]
@tobru Your dependency cycle issue occurred because we weren't namespacing the `include` of the Aptly class correctly. We've fixed this in #21 by making the includes absolute. Are you able to re-base this branch against master now? This should, hopefully, leave you with a passing build and no dependency cycles.

**Comment by tobru** [comment]
Thanks for looking into it. I will do the rebasing and retesting next week.

**Comment by ajlanghorn** [comment]
How's this going, @tobru? It would be good to get some traction on this, else we might need to close it off - at least temporarily - soon.

**Comment by tobru** [comment]
Here we are

**Comment by ajlanghorn** [comment]
Thanks @tobru; are you still getting that depcycle despite rebasing against master?

**Comment by tobru** [comment]
It looks like. To rule out some rebasing/merging fails, I did a fresh clone of your master and added just the following to init.pp:
```
diff --git a/manifests/init.pp b/manifests/init.pp
index 3af12f0..389bb2c 100644
--- a/manifests/init.pp
+++ b/manifests/init.pp
@@ -33,6 +33,7 @@ class aptly (
   $repo = true,
   $key_server = undef,
   $user = 'root',
+  $aptly_repos = {},
 ) {
 
   validate_hash($config)
@@ -61,4 +62,6 @@ class aptly (
     ensure  => file,
     content => inline_template("<%= @config.to_pson %>\n"),
   }
+
+  create_resources('::aptly::repo', $aptly_repos)
 }
```

And I'm getting the dependency cycle:
```
Error: Could not apply complete catalog: Found 1 dependency cycle:
(Exec[aptly_repo_create-my_example_repo] => Aptly::Repo[my_example_repo] => Class[Aptly] => Exec[aptly_repo_create-my_example_repo])
```

The hiera YAML looks like yours:
```
classes:
  - aptly

aptly::aptly_repos:
  my_example_repo:
    component: foo
```

**Comment by tobru** [comment]
The dependency cycle somehow makes sense: We are including the main class and pass a hash of repos to it which then calls the function `create_resources` to instantiate the `aptly::repo` defined type. This defined type in turn requires the main class to be already there. And now we have the dependency cycle. I propose to do the following change, which eliminates the dependency cycle:

```
diff --git a/manifests/repo.pp b/manifests/repo.pp
index fa5e674..c2a86ea 100644
--- a/manifests/repo.pp
+++ b/manifests/repo.pp
@@ -31,7 +31,7 @@ define aptly::repo(
     unless  => "${aptly_cmd} show ${title} >/dev/null",
     user    => $::aptly::user,
     require => [
-      Class['::aptly'],
+      Package['aptly'],
     ],
   }
 }
```

What do you think?

**Comment by tobru** [comment]
Ping @ajlanghorn: What do you think about the proposed change? This could help to finish this PR

**Comment by rjw1** [comment]
@tobru I think thats a plausible change.

**Comment by tobru** [comment]
Thanks @rjw1 for the answer.

Change committed and tested with the following data:

```
aptly::aptly_repos:
  my_example_repo:
    component: foo
aptly::aptly_mirrors:
  my_example_mirror:
    location: 'http://repo.aptly.info'
    key: '2A194991'
    release: 'squeeze'
```
outcome:
```
Notice: /Stage[main]/Aptly/Aptly::Repo[my_example_repo]/Exec[aptly_repo_create-my_example_repo]/returns: executed successfully
Notice: /Stage[main]/Aptly/Aptly::Mirror[my_example_mirror]/Exec[aptly_mirror_create-my_example_mirror]/returns: executed successfully
```

Dependency cycle: gone :clap:

**Comment by rjw1** [comment]
Closing since this has been replaced by #22 
@tobru I cleaned up the commit history and added one more require in #22
Can you confirm you're happy with being the owner of the commit.

**Pull request closed without merging**